### PR TITLE
Add logic to fail DPA reconcile on wrong BSL secret key

### PR DESCRIPTION
**Pull request by sbahar619**
This PR addresses:
https://issues.redhat.com/browse/OADP-2940

### CODE REVIEW

**Comment by sbahar619** [comment]
Verified:
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/cam/job/oadp-pipeline/5903/

**Comment by kaovilai** [comment]
Just need to fix unit tests for the new validator :).

I can push a fix here if assistance is needed. LMK.

File: controllers/bsl.go [hunk 1]
```diff
@@ -102,12 +102,29 @@ func (r *DPAReconciler) ReconcileBackupStorageLocations(log logr.Logger) (bool,
 		// 	 2. <namespace>.dataprotectionapplication: <name>
 		// which in turn will be used in th elabel handler to trigger the reconciliation loop
 
-		secretName, _ := r.getSecretNameAndKeyforBackupLocation(bslSpec)
+		secretName, secretKey := r.getSecretNameAndKeyforBackupLocation(bslSpec)
```
Comment (shubham-pampattiwar) [inline controllers/bsl.go:105]: Maybe we should move this to `ValidateDataProtectionCR` [function](https://github.com/openshift/oadp-operator/blob/22d063a852175a5d89835d856f450c9ae596965c/controllers/dpa_controller.go#L95) ?

**Review by shubham-pampattiwar — COMMENTED** [review]
Added a suggestion, plus yeah unit tests need to be fixed

File: controllers/bsl.go (hunk 1 above)
Comment (kaovilai) [reply controllers/bsl.go:105]: Could do that too

Comment (sbahar619) [reply controllers/bsl.go]: Done and verified.

**Comment by sbahar619** [comment]
> Thank you @sbahar619 ! Added a suggestion, plus yeah unit tests need to be fixed

Thanks @shubham-pampattiwar , I applied the suggestion and verified.

**Comment by sbahar619** [comment]
> Just need to fix unit tests for the new validator :).
> 
> I can push a fix here if assistance is needed. LMK.

@kaovilai 
I pushed some unit tests fixes without verifying, please review and run the tests.

**Comment by sbahar619** [comment]
> Just need to fix unit tests for the new validator :).
> 
> I can push a fix here if assistance is needed. LMK.

Thanks @kaovilai 
I pushed fixes for the unit tests and all the tests pass now.
Please review my changes and verify that I am not destroy the intended tests checks.
Also, one of the unit test actually helped me to realized that my PR  missing some coverage on Velero "no-secret" flag, so I added that logic too.

**Comment by sbahar619** [comment]
> @sbahar619 unfortunately we have a flake in our CI for E2E tests. No need to run `retest` command

Opps, ok thanks, got it.

File: controllers/validator.go [hunk 2]
```diff
@@ -55,6 +55,31 @@ func (r *DPAReconciler) ValidateDataProtectionCR(log logr.Logger) (bool, error)
 		if location.CloudStorage != nil && location.CloudStorage.Prefix == "" && dpa.BackupImages() {
 			return false, errors.New("BackupLocation must have cloud storage prefix when backupImages is not set to false")
 		}
+
+		// Check if the Velero flag 'no-secret' is not specified
+		if !strings.Contains(strings.Join(dpa.Spec.Configuration.Velero.FeatureFlags, " "), "no-secret") {
```
Comment (kaovilai) [inline controllers/validator.go:60]:
```suggestion
		if !strings.Contains(strings.Join(dpa.Spec.Configuration.Velero.FeatureFlags, " "), "no-secret") || !dpa.Spec.Configuration.Velero.NoDefaultBackupLocation {
```

File: controllers/validator.go [hunk 3]
```diff
@@ -55,6 +55,31 @@ func (r *DPAReconciler) ValidateDataProtectionCR(log logr.Logger) (bool, error)
...
+			for currentKey := range(bslSecret.Data) {
+				if secretKey == currentKey {
+					foundKey = true
+					break
+				}
+			}
+
+			if !foundKey {
+				return false, fmt.Errorf("Secret name %s is missing data for key %s", secretName, secretKey)
+			}
```
Comment (kaovilai) [inline controllers/validator.go:81]:
```suggestion
			data, foundKey := bslSecret.Data[secretKey]

			if !foundKey || len(data) == 0 {
				return false, fmt.Errorf("Secret name %s is missing data for key %s", secretName, secretKey)
			}
```

**Review by kaovilai — CHANGES_REQUESTED** [review]

File: controllers/validator.go (hunk 2 above)
Comment (sbahar619) [reply controllers/validator.go]: Thanks!
Done.

File: controllers/validator.go (hunk 3 above)
Comment (sbahar619) [reply controllers/validator.go]: Much better, thanks!
Done.

**Comment by sbahar619** [comment]
@kaovilai 
I verified the last commit changes with all the added cases: noDefaultBackupLocation and featureFlags.

**Comment by kaovilai** [comment]
unit tests are green!

**Review by kaovilai — APPROVED** [review]

**Comment by kaovilai** [comment]
some flakes but the rest of e2e is failing on known issue in-cluster version error get https://github.com/vmware-tanzu/velero/issues/6952

**Comment by sbahar619** [comment]
Thanks @kaovilai !

@shubham-pampattiwar 
Can you please review?

**Review by shubham-pampattiwar — APPROVED** [review]

**Pull request merged**